packages/mdc-textfield/input/foundation.js

```diff
@@ -0,0 +1,181 @@
+ */
+class MDCTextFieldInputFoundation extends MDCFoundation {
+  /** @return enum {string} */
+  static get strings() {
+    return strings;
+  }
+
+  /**
+   * {@see MDCTextFieldInputAdapter} for typing information on parameters and return
+   * types.
+   * @return {!MDCTextFieldInputAdapter}
+   */
+  static get defaultAdapter() {
+    return /** @type {!MDCTextFieldInputAdapter} */ ({
+      registerEventHandler: () => {},
+      deregisterEventHandler: () => {},
+      getNativeInput: () => {},
+      notifyFocusAction: () => {},
+      notifyBlurAction: () => {},
+      notifyPressedAction: () => {},
+    });
+  }
+
+  /**
+   * @param {!MDCTextFieldInputAdapter=} adapter
+   */
+  constructor(adapter = /** @type {!MDCTextFieldInputAdapter} */ ({})) {
+    super(Object.assign(MDCTextFieldInputFoundation.defaultAdapter, adapter));
+
+    /** @private {boolean} */
+    this.receivedUserInput_ = false;
+    /** @private {function(): undefined} */
+    this.inputFocusHandler_ = () => this.activateFocus_();
+    /** @private {function(): undefined} */
+    this.inputBlurHandler_ = () => this.deactivateFocus_();
+    /** @private {function(): undefined} */
+    this.inputInputHandler_ = () => this.autoCompleteFocus_();
+    /** @private {function(): undefined} */
+    this.inputPressedHandler_ = () => this.notifyPressedAction_();
+  }
+
+  init() {
+    this.adapter_.registerEventHandler('focus', this.inputFocusHandler_);
+    this.adapter_.registerEventHandler('blur', this.inputBlurHandler_);
+    this.adapter_.registerEventHandler('input', this.inputInputHandler_);
+    ['mousedown', 'touchstart'].forEach((evtType) => {
+      this.adapter_.registerEventHandler(evtType, this.inputPressedHandler_);
+    });
+  }
+
+  destroy() {
+    this.adapter_.deregisterEventHandler('focus', this.inputFocusHandler_);
+    this.adapter_.deregisterEventHandler('blur', this.inputBlurHandler_);
+    this.adapter_.deregisterEventHandler('input', this.inputInputHandler_);
+    ['mousedown', 'touchstart'].forEach((evtType) => {
+      this.adapter_.deregisterEventHandler(evtType, this.inputPressedHandler_);
+    });
+  }
+
+  /**
+   * Activates the input's focus state.
+   * @private
+   */
+  activateFocus_() {
+    this.adapter_.notifyFocusAction();
+  }
+
+  /**
+   * Activates the input's focus state in cases when the input value
+   * changes without user input (e.g. programatically).
+   * @private
+   */
+  autoCompleteFocus_() {
+    if (!this.receivedUserInput_) {
+      this.activateFocus_();
+    }
+  }
+
+  /**
+   * Deactives the input's focus state.
+   * @private
+   */
+  deactivateFocus_() {
+    const input = this.getNativeInput_();
+
+    if (!input.value && !this.isBadInput()) {
+      this.receivedUserInput_ = false;
+    }
+
+    this.adapter_.notifyBlurAction();
+  }
+
+  notifyPressedAction_() {
+    this.adapter_.notifyPressedAction();
+  }
+
+  /**
+   * @return {boolean} True if the input fails validity checks.
+   */
+  isBadInput() {
```

Comment:
I realize this code existed prior to this split, but...I'm confused why this exists. What is this doing distinct from checkValidity? cc @amsheehan

Reply:
Lets not architecture in this PR. This PR is just refactoring. If it illuminates some problems with the old MDCTextFieldFoundation code...great! But file issues and we will resolve in another PR.